File: Data Structures & Algorithms/subsets/submission-5.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        res = []

        def generateSet(i, curr):
            if i >= len(nums):
                res.append(curr[:])
                return

            curr.append(nums[i])
            generateSet(i + 1, curr)

            curr.pop()
            generateSet(i + 1, curr)

        generateSet(0, [])
        return res

File: Data Structures & Algorithms/combinations/submission-5.ts
class Solution {
    /**
     * @param {number} n
     * @param {number} k
     * @return {number[][]}
     */
    combine(n: number, k: number): number[][] {
        const res: number[][]  = []

        const backtracking = (i: number, curr: number[]): void => {
            if (curr.length === k) {
                res.push([...curr])
                return
            }

            if (i > n) {
                return
            }

            curr.push(i)
            backtracking(i + 1, curr)
            curr.pop()
            backtracking(i + 1, curr)
        }

        backtracking(1, [])
        return res
    }
}
